pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/InternalInterfaces.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+ */
+
+package net.sourceforge.pmd.lang.java.ast;
+
+
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+/**
+ * Those are some interfaces that are not published, but are used to keep
+ * uniform names on related concepts. Maybe it makes sense to publish some of
+ * them at some point.
+ */
+@SuppressWarnings("PMD.MissingStaticMethodInNonInstantiatableClass")
+final class InternalInterfaces {
+
+    private InternalInterfaces() {
+        // utility class
+    }
+
+    interface OperatorLike {
+
+
+        /**
+         * Returns the token used to represent the type in source
+         * code, e.g. {@code "+"} or {@code "*"}.
+         */
+        String getToken();
+
+    }
+
+    /** Just to share the method names. */
+    interface BinaryExpressionLike extends ASTExpression {
+
+        /** Returns the left-hand-side operand. */
+        default ASTExpression getLhs() {
+            return (ASTExpression) jjtGetChild(0);
+        }
+
+
+        /** Returns the right-hand side operand. */
+        default ASTExpression getRhs() {
+            return (ASTExpression) jjtGetChild(1);
+        }
+
+
+        /** Returns the operator. */
+        OperatorLike getOperator();
+    }
+
+    /**
+     * Node that may be qualified by an expression, e.g. an instance method call or
+     * inner class constructor invocation.
+     */
+    interface ASTQualifiableExpression extends ASTExpression {
+
+        /**
+         * Returns the expression to the left of the "." if it exists.
+         * This may be a {@link ASTTypeExpression type expression}, or
+         * an {@link ASTAmbiguousName ambiguous name}.
+         */
+        @Nullable
+        default ASTExpression getQualifier() {
+            return AstImplUtil.getChildAs(this, 0, ASTExpression.class);
+        }
+    }
+
+    /**
+     * A node that has a single type of children, but can have no children.
+     *
+     * Package private, the methods are exposed nevertheless.
+     *
+     * @author Clément Fournier
+     */
+    interface JSingleChildNode<T extends JavaNode> extends JavaNode {
```

Comment:
From the pure name of the method, I expected that the node has only one child. But it has children of only one type... but I don't really know a better name. maybe "JSingleChildTypeNode"? "JSingleTypeNode"?

Reply:
Hmm yes, quite honestly I think we should remove this interface though. The methods usually have a semantic counterpart that should be preferred, eg InfixExpression has getLhs/getRhs. So it's not extremely useful.

The main problem, is that this interface makes it harder to modify existing nodes, eg when we want to add support for a new language feature. For example, #1980 made the receiver parameter different from FormalParameter, and ASTFormalParameter**s** needed to unimplement the interface. 

One possible area where this may cause problems, is if eg at some point we make all comments available as nodes. Then eg `jjtGetChild` would need to always return `JavaNode` anyway.